internal/catalogmetadata/client/client.go

```diff
@@ -43,66 +47,93 @@ type Client struct {
 	fetcher Fetcher
 }
 
-func (c *Client) Bundles(ctx context.Context) ([]*catalogmetadata.Bundle, error) {
+func (c *Client) Bundles(ctx context.Context, packageName string) ([]*catalogmetadata.Bundle, error) {
 	var allBundles []*catalogmetadata.Bundle
 
 	var catalogList catalogd.ClusterCatalogList
 	if err := c.cl.List(ctx, &catalogList); err != nil {
 		return nil, err
 	}
+
+	var errs []error
 	for _, catalog := range catalogList.Items {
-		// if the catalog has not been successfully unpacked, skip it
+		// if the catalog has not been successfully unpacked, report an error. This ensures that our
+		// reconciles are deterministic and wait for all desired catalogs to be ready.
 		if !meta.IsStatusConditionPresentAndEqual(catalog.Status.Conditions, catalogd.TypeUnpacked, metav1.ConditionTrue) {
+			errs = append(errs, fmt.Errorf("catalog %q is not unpacked", catalog.Name))
```

Comment:
This is a change in semantics. If we ignore catalogs that are not unpacked, users could run into issues if the same package exists in two different catalogs and they are expecting a specific result.

This is similar to the behavior in OLMv0, and there have been some complaints about it. For example, "I want to install CNV, I don't care if the community catalog is having issues". In order to solve this problem, I think we need to have a knob on the ClusterExtension API to allow users to select which ClusterCatalogs to resolve from (likely using a label selector)

Reply:
Do we have this captured as an issue somewhere or do you anticipate this falling under the umbrella of https://github.com/operator-framework/operator-controller/issues/740?

Reply:
I am going to make a separate issue for this. Just wanted to get consensus on this approach before I did.

Reply:
https://github.com/operator-framework/operator-controller/issues/968